File: 870-magic-squares-in-grid/magic-squares-in-grid.cpp
class Solution {
public:
    bool solve(int x, int y, vector<vector<int>>& grid) {
        vector<int> row(3), col(3);
        vector<int> vis(16, 0);  // Initialize with 0
        int d1 = 0, d2 = 0;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                int num = grid[x + i][y + j];
                vis[num]++;
                row[i] += num;
                col[j] += num;
                if (i == j) d1 += num;
                if (i + j == 2) d2 += num;
            }
        }
        // Early return if diagonals don't match
        if (d1 != d2) return false;
        // Check if all rows and columns have the same sum as the diagonals
        for (int i = 0; i < 3; i++) {
            if (row[i] != d1 || col[i] != d1) return false;
        }
        // Check if numbers 1 through 9 are all present exactly once
        for (int i = 1; i <= 9; i++) {
            if (vis[i] != 1) return false;
        }
        return true;
    }
    int numMagicSquaresInside(vector<vector<int>>& grid) {
        int cnt = 0;

        // Ensure grid is large enough for a 3x3 subgrid
        if (grid.size() < 3 || grid[0].size() < 3) return 0;

        // Iterate over each possible starting position for a 3x3 subgrid
        for (int i = 0; i <= grid.size() - 3; i++) {
            for (int j = 0; j <= grid[0].size() - 3; j++) {
                // Check if the subgrid starting at (i, j) is a magic square
                if (solve(i, j, grid)) cnt++;
            }
        }
        return cnt;
    }
};

File: 870-magic-squares-in-grid/magic-squares-in-grid.py
import numpy as np
from typing import List
class Solution:
    def is_magic_square(self, subgrid: np.ndarray) -> bool:
        # Check if all numbers from 1 to 9 are present
        if set(subgrid.flatten()) != set(range(1, 10)):
            return False
        # Calculate the sum of the diagonals
        d1 = np.sum(np.diag(subgrid))
        d2 = np.sum(np.diag(np.fliplr(subgrid)))
        # If diagonals are not equal, return False
        if d1 != d2:
            return False
        # Check if all rows and columns have the same sum as the diagonals
        if not all(np.sum(subgrid, axis=0) == d1) or not all(np.sum(subgrid, axis=1) == d1):
            return False
        return True
    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        grid = np.array(grid)
        rows, cols = grid.shape
        count = 0
        for i in range(rows - 2):
            for j in range(cols - 2):
                subgrid = grid[i:i+3, j:j+3]
                if self.is_magic_square(subgrid):
                    count += 1
        return count
